crdt/sequence.py

```diff
@@ -0,0 +1,247 @@
+
+class OpId():
+    """
+    An OpId represents a unique identifier for an Operation.
+    """
+
+    def __init__(self, name, id):
+        self.name = name
+        self.id = id
+
+    def compare(self, other):
+        if self.id < other.id:
+            return -1
+        elif self.id > other.id:
+            return 1
+        else:
+            # Conflicts are resolved arbitrarily by comparing the owner IDs
+            # TODO: We could handle "forks" of the operation log and allow for multiple
+            # histories to be valid.
+            if self.name < other.name:
+                return -1
+            elif self.name > other.name:
+                return 1
+            else:
+                return 0
+
+    def is_earlier(self, other):
+        """
+        Returns true if this OpId is earlier than the other OpId.
+        """
+        return self.compare(other) < 0
+
+    def __eq__(self, other):
+        if not isinstance(other, OpId):
+            return False
+        return self.name == other.name and self.id == other.id
+
+    def print(self):
+        """
+        Prints the OpId to stdout.
+        """
+        return "name: {}, id: {}".format(self.name, self.id)
+
+class OperationType(Enum):
+    """
+    Enum for operations.
+    """
+    INSERT = 0
+    REMOVE = 1
+
+class Operation():
+    """
+    An Operation represents a point-in-time change to a Sequence. Every Operation
+    has a globally-unique ID and references a previous Operation, enabling an order of
+    Operations to be imposed.
+    """
+
+    def __init__(self, owner, action, target, payload=None):
+        if action not in OperationType:
+            raise ValueError("Invalid operation type")
+        self.owner = owner
+        self.action = action
+        self.target = target
+        self.payload = payload
+
+    def do(self, objects):
+        """
+        Applies this Operation to an ordered list of objects.
+        """
+        # Iterate through the sequence to find the index of the target object
+        stomp = False
+        if self.target is None:
+            # Special case: Append at the end
+            index = len(objects)
+            while index > 0:
+                op = objects[index-1].operation
+                if op.target is None and not self.owner.is_earlier(op.owner):
+                    # We found another operation older than us, so stop iteration.
+                    break
+                index -= 1
+        else:
+            # Normal case: Find the object to insert before
+            index = 0
+            while index < len(objects):
+                op = objects[index].operation
+                if op.target == self.target and self.owner.is_earlier(op.owner):
+                    # We found another operation older than us, so stop iteration.
+                    break
+                elif op == self.target:
+                    # We found the target index
+                    break
+                index += 1
```

Comment:
This is the most important piece of code; it handles the merging of operations into the object list so that we can maintain the current state of the sequence. Since we are inserting operations after the fact (due to getting unseen operations from the remote), this requires a scan of the object list which in the worst case we have to scan the entire operation set. There is room for optimization here; maybe we could also discard tombstoned operations periodically so the search space doesn't get unwieldy.